### ISSUE TITLE
Performance of Tracer seems to be a lot slower on current master

### ISSUE THREAD

**Issue by sean797** [FrostyX/tracer#86]
Performance using the API isn't too bad, but the shell command is a lot slower 👎 

```sh
$ sync && echo 3 > /proc/sys/vm/drop_caches
$ time tracer -ea (0.6.12)
<snip>

real	0m30.557s
user	0m21.941s
sys	0m8.268s
$ cd git/tracer
$ sync && echo 3 > /proc/sys/vm/drop_caches
$ time ./bin/tracer.py -ea  (0.6.13)
<snip>

real	4m16.887s
user	2m25.921s
sys	1m44.147s
```

Using the API

```sh
$ sync && echo 3 > /proc/sys/vm/drop_caches
$ time python tt.py  (0.6.12)
<snip>

real	0m8.003s
user	0m4.773s
sys	0m2.945s
$ cd git/tracer
$ sync && echo 3 > /proc/sys/vm/drop_caches
$ time python tt.py  (0.6.13)
<snip>

real	0m53.491s
user	0m30.489s
sys	0m21.493s
```
tt.py contents:
```python
from tracer.query import Query
q = Query()
apps = {}
for app in q.affected_applications().get():
   apps[app.name] = { "helper": app.helper, "type": app.type}
print apps
```

**Comment by AlliedEnvy** [comment]
Seconding this issue. I've had to disable the dnf plugin because it just takes too long.

I bisected the start of the problem to 51273d2bb5054ffd988e68ab9b4c30fd221f9053. However, after ab841376cb624b16b0fa59923ff4d149eb2514fe, none of the changes I've tried seem to help.

**Comment by sean797** [comment]
@AlliedEnvy interesting... I assumed the issue was https://github.com/FrostyX/tracer/commit/ee1c7f70e4d195138f25c1479c8a6729f4e36d50

**Comment by elyscape** [comment]
I ran the Python profiler on the code on a system with about 100 running processes:
```shell
# run the profile
sudo python3 -m cProfile -o profile.prf /usr/bin/tracer -ea
# launch profile analyzer
python3 -m pstats profile.prf
```
Something stood out to me. Some context for the following: total time is defined as the time spent executing a function's actual code code (i.e. exclusive of time spent inside functions that it calls), whereas cumulative time is defined as the time spent inside a function (i.e. inclusive of time spent inside functions that it calls).

There is a lot of time spent inside psutil.
```
profile.prf% sort cumtime
profile.prf% stats 10
Fri Apr  6 13:28:52 2018    profile.prf

         67514569 function calls (67510246 primitive calls) in 131.461 seconds

   Ordered by: cumulative time
   List reduced from 1680 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    258/1    0.016    0.000  131.461  131.461 {built-in method builtins.exec}
        1    0.000    0.000  131.461  131.461 /usr/bin/tracer:20(<module>)
        1    0.000    0.000  131.299  131.299 /usr/lib/python3.6/site-packages/tracer/main.py:30(run)
        1    0.000    0.000  131.299  131.299 /usr/lib/python3.6/site-packages/tracer/resources/router.py:36(dispatch)
        1    0.000    0.000  128.689  128.689 /usr/lib/python3.6/site-packages/tracer/controllers/default.py:66(render)
     4608    0.151    0.000  121.040    0.026 /usr/lib/python3.6/site-packages/tracer/resources/applications.py:253(instances)
        1    0.000    0.000  114.758  114.758 /usr/lib/python3.6/site-packages/tracer/views/default.py:15(render)
      566    0.003    0.000  113.386    0.200 /usr/lib/python3.6/site-packages/tracer/resources/applications.py:198(type)
      945    0.044    0.000  109.488    0.116 /usr/lib/python3.6/site-packages/tracer/resources/applications.py:266(name)
  3614310    3.391    0.000  105.086    0.000 /usr/lib64/python3.6/site-packages/psutil/_pslinux.py:1387(wrapper)
```
A lot of that time time is spent inside the internal function `_parse_stat_file`, which is called an absurd 3,296,071 times:
```
profile.prf% sort tottime
profile.prf% stats 10
Fri Apr  6 13:28:52 2018    profile.prf

         67514569 function calls (67510246 primitive calls) in 131.461 seconds

   Ordered by: internal time
   List reduced from 1680 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  3614325   31.300    0.000   31.664    0.000 {built-in method io.open}
  3296071   23.664    0.000   80.243    0.000 /usr/lib64/python3.6/site-packages/psutil/_pslinux.py:1419(_parse_stat_file)
  3296170   21.187    0.000   21.187    0.000 {method 'read' of '_io.BufferedReader' objects}
  3446992    4.654    0.000    4.654    0.000 {method 'split' of 'bytes' objects}
  2819968    3.894    0.000   92.799    0.000 /usr/lib64/python3.6/site-packages/psutil/__init__.py:599(name)
3296212/3296174    3.425    0.000   83.709    0.000 /usr/lib64/python3.6/site-packages/psutil/_common.py:334(wrapper)
  3614310    3.391    0.000  105.086    0.000 /usr/lib64/python3.6/site-packages/psutil/_pslinux.py:1387(wrapper)
  2819968    3.367    0.000   77.231    0.000 /usr/lib64/python3.6/site-packages/psutil/_pslinux.py:1464(name)
     3221    3.152    0.001    3.984    0.001 {method 'Parse' of 'pyexpat.xmlparser' objects}
   318136    2.514    0.000   10.302    0.000 /usr/lib64/python3.6/site-packages/psutil/_pslinux.py:1491(cmdline)
```
The reason it's called so often is that psutil does not cache information unless _specifically advised to do so_. This is done by running within a `Process.oneshot()` context. As an example:
```python
import os
import psutil
import time

def print_ctx_switches(proc):
    print(proc.num_ctx_switches())
    # force a context switch
    time.sleep(1)
    print(proc.num_ctx_switches())

myproc = psutil.Process(os.getpid())

print('Regular execution:')
print_ctx_switches(myproc)

print('Inside oneshot:')
with myproc.oneshot():
    print_ctx_switches(myproc)
```
Running this script yields the following:
``` shellsession
$ python3 ./proc_cache.py
Regular execution:
pctxsw(voluntary=0, involuntary=21)
pctxsw(voluntary=1, involuntary=23)
Inside oneshot:
pctxsw(voluntary=1, involuntary=27)
pctxsw(voluntary=1, involuntary=27)
```
Since tracer is just using the various methods of `Process` regularly without wrapping them inside `Process.oneshot()`, psutil ends up reading and parsing various process stat files a ridiculous number of times.

Unfortunately, `Process.oneshot()` was added to psutil in version 5.0.0, meaning that we can't use it and retain backwards compatibility with older versions of the library (as are found on CentOS 7, Debian pre-9, and Ubuntu pre-17.04). The good news is that there's also a `Process.as_dict()` method, which unsurprisingly returns a dict of the data about the process, and which has been available since at least version 0.7.0. On psutil 5.0.0+, it uses `Process.oneshot()` internally, but it's available in earlier versions as well and, more importantly, it would allow us to only have the overhead of parsing the stat files a constant number of times per process.

**Comment by elyscape** [comment]
Some additional info:
```
profile.prf% strip
profile.prf% sort tottime
profile.prf% callers _parse_stat_file
   Ordered by: internal time
   List reduced from 1669 to 1 due to restriction <'_parse_stat_file'>

Function                            was called by...
                                        ncalls  tottime  cumtime
_pslinux.py:1419(_parse_stat_file)  <- 3296071   23.664   80.243  _common.py:334(wrapper)


profile.prf% callers 334.wrapper
   Ordered by: internal time
   List reduced from 1669 to 1 due to restriction <'334.wrapper'>

Function                 was called by...
                             ncalls  tottime  cumtime
_common.py:334(wrapper)  <- 2819968    2.912   70.553  _pslinux.py:1464(name)
                                386    0.001    0.013  _pslinux.py:1516(terminal)
                             475679    0.512   13.101  _pslinux.py:1570(create_time)
                                103    0.000    0.041  _pslinux.py:1635(memory_maps)
                                 38    0.000    0.002  _pslinux.py:1959(ppid)
                                 38    0.000    0.002  processes.py:100(_attr)
```
So what's going on is primarily that `Process.name()` is getting called a whole bunch of times and that's resulting in the stat files getting read and parsed nearly 3 million times.

**Comment by elyscape** [comment]
Additional followup: doing some naive implementation that uses `psutil.Process.as_dict()` helps some but reveals that new process objects are still spun up a _lot_, like in [the reimplementations of `Process.parent()` and `Processes.children()`][reimpl], and each time that happens we need to parse the process's memory map. Further wrapping `Process` in a metaclass that caches the resulting objects and only creates new objects for new processes ([see here for example code][metaclass]) helps immeasurably.

To sum up, we should:
1. Call `psutil.Process.as_dict()` when a `ProcessesWrapper` object is instantiated and use that to provide the other values
1. Add a caching metaclass wrapper to `Process`

[reimpl]: https://github.com/FrostyX/tracer/blob/e706cd499415a3af2213772dabd28792b00cb2ec/tracer/resources/processes.py#L149-L169
[metaclass]: https://stackoverflow.com/a/33458129
